sim/dex.ts

```diff
@@ -486,22 +486,30 @@ class ModdedDex {
 			// @ts-ignore
 			return effect;
 		}
+		return this.getEffectByID(id, effect);
+	}
+
+	getEffectByID(id: ID, effect?: Effect | Move): PureEffect {
+		if (!id) return nullEffect;
+
+		if (!effect) effect = this.effectCache.get(id);
+		if (effect) return effect as PureEffect;
 
 		let found;
 		if (this.data.Formats.hasOwnProperty(id)) {
-			effect = new Data.Format({name}, this.data.Formats[id]);
+			effect = new Data.Format({name: id}, this.data.Formats[id]);
 		} else if (this.data.Statuses.hasOwnProperty(id)) {
-			effect = new Data.PureEffect({name}, this.data.Statuses[id]);
+			effect = new Data.PureEffect({name: id}, this.data.Statuses[id]);
 		} else if ((this.data.Movedex.hasOwnProperty(id) && (found = this.data.Movedex[id]).effect) ||
 							 (this.data.Abilities.hasOwnProperty(id) && (found = this.data.Abilities[id]).effect) ||
 							 (this.data.Items.hasOwnProperty(id) && (found = this.data.Items[id]).effect)) {
-			effect = new Data.PureEffect({name: found.name || name}, found.effect!);
+			effect = new Data.PureEffect({name: found.name || id}, found.effect!);
 		} else if (id === 'recoil') {
-			effect = new Data.PureEffect({name: 'Recoil', effectType: 'Recoil'});
+			effect = new Data.PureEffect({id, name: 'Recoil', effectType: 'Recoil'});
 		} else if (id === 'drain') {
-			effect = new Data.PureEffect({name: 'Drain', effectType: 'Drain'});
+			effect = new Data.PureEffect({id, name: 'Drain', effectType: 'Drain'});
 		} else {
-			effect = new Data.PureEffect({name, exists: false});
+			effect = new Data.PureEffect({id, name: id, exists: false});
```

Comment:
This should probably throw instead of creating a placeholder.

Reply:
I don't know how I feel about how PS creates placeholders. I don't _think_ there are any situations we currently use it that aren't bugs? But in any case, if the placeholder doesn't even have a name, that should definitely throw.

Reply:
Well, `drain` etc are currently placeholders. We'd have to make sure to make actual entries in `statuses.js` for them.

Reply:
> I named `toId` based on precedents set by DOM APIs with stuff like `getElementById`. But PS conventions elsewhere seem to think `toID` is better. I'm fine with `ID` instead of `Id` here.

I'm fine with either `Id` or `ID`, I just don't like the inconsistency of having a type called `ID` and methods like `fooId`. To that end, I'd be happy to rename `toId` -> `toID` in the codebase(s)...

> I've also considered `ID()`, by analogy to how e.g. `String()` is a casting function. TypeScript _does_ support that, and it would be consistent with our precedent that globals are allcaps. 

Since when are our globals all caps??? We have a ton of globals, none of which are ALLCAPS from what I can tell?

> Judging by votes on `Rooms()` vs `Rooms.get()`, you might all prefer `ID.get()`, which, ew.

*Oh, you did see that PR then... ;)*

I prefer `toID`. If `ID` were a class, then yeah, probably `new ID()`/`ID.create`/`ID.get`. Though this might be the one single place where I'd be OK with `ID()` given its ubiquity and how its more analogous to a primitive than something like `Rooms` or `Users` (or, `TeamValidator`...)

Reply:
I have a note to myself to talk to you about the `{exists: false}` pattern - why don't we just return `undefined`? Obviously, our code doesnt want to need to check for `undefined`, but creating bulky objects (`PureEffect` has a good 20+ fields) because we don't want to check `undefined` (which the compiler can help with statically) seems really inefficient.

Reply:
The biggest reason we don't just return `undefined` is because https://github.com/tc39/proposal-optional-chaining is _still_ Stage 1 after approximately eight years of literally everyone asking for it.

Reply:
Honestly, we even considered making placeholder `Pokemon` objects, for similar reasons in #697.

The only reason we didn't was because a researcher pointed out that pokemon slots can never actually be completely empty.

Reply:
This doesn't seem resolved – you've neither added a throw nor said that you think it shouldn't.

Reply:
Ohhh, I see, you consider that a separate change. I originally thought this was a regression, but I guess in practice it isn't.

Reply:
I still consider this a distinct change, though! I'm recommending throwing, instead of returning `null`, because there are significantly fewer `getEffectByID` calls, and it's significantly easier to audit every call and note that they'd all be bugs if not found.

Reply:
I guess that can be a separate change, though